spec/Classes.tex

```diff
@@ -663,73 +604,456 @@ \subsection{User-Defined Constructors}
+based on the type of those fields (\rsec{Default_Values_For_Types}).
+\end{chapelexample}
+
+In order to reduce ambiguity and to ensure a well-defined order for
+side-effects, the language requires that all fields be initialized in field
+declaration order. This applies regardless of whether field initializations are
+omitted from the initializer body. If fields are initialized out of order, a
+compile-time error will be issued.
+
+\begin{chapelexample}{fieldsOutOfOrder.chpl}
+In the following code:
+\begin{chapel}
+class Point3D {
+  var x = 1.0;
+  var y = 1.0;
+  var z = 1.0;
+
+  proc init(x: real) {
+    this.x = x;
+    // compiler inserts "this.y = 1.0;"
+    this.z = y * 2.0;
+  }
+
+  proc init(x: real, y: real, z: real) {
+    this.x = x;
+    this.z = z;
+    this.y = y; // Error!
+  }
+}
+
+var A = new Point3D(1.0);
+var B = new Point3D(1.0, 2.0, 3.0);
+\end{chapel}
+\begin{chapelcompopts}
+--no-warnings
+\end{chapelcompopts}
+\begin{chapeloutput}
+fieldsOutOfOrder.chpl:12: In initializer:
+fieldsOutOfOrder.chpl:17: error: Field "y" initialized out of order
+fieldsOutOfOrder.chpl:17: note: initialization of fields before .init() call must be in field declaration order
+\end{chapeloutput}
+
+The first initializer leverages the well-defined order of omitted field
+initialization to use the default value of field \chpl{y} in order to
+explicitly initialize field \chpl{z}.
+
+The second initializer initializes field \chpl{z} before field \chpl{y},
+causing a compile-time error to be issued.
+
+\end{chapelexample}
+
+\begin{rationale}
+Without this rule the compiler could insert default initialization for field
+\chpl{y} before \chpl{z} is explicitly initialized. The following statement
+would then be {\em assignment} to field \chpl{y}, despite appearing to be
+initialization. This subtle difference may be confusing and surprising, and
+is avoided by requiring fields to be initialized in field declaration order.
+\end{rationale}
+
+\subsubsection{The complete Method}
+\label{The_Complete_Method}
+\index{classes!initializers!user-defined!complete}
+
+As the initializer makes progress, the class instance is incrementally
+initialized. In order to prevent usage of uninitialized memory, there are
+restrictions on usage of the class instance before it is fully initialized:
+
+\begin{itemize}
+\item Methods may not be invoked on partially-initialized instances
+\item \chpl{this} may not be passed to functions while partially-initialized
+\end{itemize}
+
+These rules allow all methods and functions to assume that class instances
+have been initialized, provided their value is not \chpl{nil}.
+
+\begin{rationale}
+  The compiler could conceivably attempt to analyze methods and functions to
+  determine which fields are used, and selectively allow method calls on
+  partially-initialized class instances. A simpler rule, which is easier to
+  remember and implement, is to forbid method calls on partially-initialized
+  instances.
```

Comment:
I might end this paragraph with explicitly repeating which strategy is recommended/described by the spec.